File: built-in plugins dsp/mb-vocal-comp.cpp
/**
 * MB Vocal Compressor
 * Category : effect
 * Type     : vocal
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Optimized vocal dynamics
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_VOCAL_COMP_H
#define MB_VOCAL_COMP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbVocalComp : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-vocal-comp";
    static constexpr const char* PLUGIN_NAME    = "MB Vocal Compressor";
    static constexpr const char* PLUGIN_TYPE    = "vocal";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -18f;  // range [-60, 0]
    float ratio = 4f;  // range [1, 20]
    float attack = 5f;  // range [0.1, 50]
    float release = 80f;  // range [10, 500]
    };

    MbVocalComp() = default;
    ~MbVocalComp() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -60f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 20f);
        params.attack = std::clamp(params.attack, 0.1f, 50f);
        params.release = std::clamp(params.release, 10f, 500f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Vocal Compressor
        return input;
    }
};

#endif // MB_VOCAL_COMP_H

File: built-in plugins dsp/mb-vocal-comp.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbVocalCompPlugin: PluginDefinition = { id: 'mb-vocal-comp', slug: 'mb-vocal-comp', name: 'MB Vocal Compressor', category: 'effect', type: 'vocal', version: '1.0.0', description: 'Optimized vocal dynamics', author: 'Max Booster', grade: 'A', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -18, minValue: -60, maxValue: 0, automatable: true }, { id: 'ratio', name: 'Ratio', type: 'float', defaultValue: 4, minValue: 1, maxValue: 20, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 5, minValue: 0.1, maxValue: 50, automatable: true }, { id: 'release', name: 'Release', type: 'float', defaultValue: 80, minValue: 10, maxValue: 500, automatable: true }], defaultPreset: { threshold: -18, ratio: 4, attack: 5, release: 80 } };

export default MbVocalCompPlugin;
